### PR TITLE
pairs & map_pairs generators, with_key & index_by transformers

### PR DESCRIPTION
**Pull request by printercu**
Using `iter` is dangerous on sparse arrays: it produces completely different generators based on `#obj > 0` condition, while the result of this condition is unpredictable for sparse arrays in lua.

Also `iter` is the only way to create map-generator, which passes key and value to filter/transform function. But it's works wrong when input value appears to have integer keys: in unpredictable way `iter` can return array-iterator or map-iterator.

This PR introduces `fun.pairs` and `fun.map_pairs` functions to create generators in explicit way. Both works the same way as `fun.iter(pairs(...))` but the second one returns map-generator.

Additional `with_key`, `index_by` methods take any generator and return map-generator for it. `with_key` uses `state` as key value which works fine for the most cases, and `index_by` takes function to generate key value with. This makes it possible to generate map from table with luafun only:

```lua
input = {{'a', 1}, {'b', 2}}

fun.iter(input):with_key():map(function(_, v) return v[1], v[2] end):tomap()
-- > {a = 1, b = 2}
fun.iter(input):index_by(function(v) return v[1] end):tomap()
-- > {a = {'a', 1}, b = {'b', 2}}
```

### CODE REVIEW

File: tests/runtest [hunk 1]
```diff
@@ -1,7 +1,8 @@
 #!/usr/bin/env lua
 
 package.path = "../?.lua;"..package.path
-require "fun" ()
+fun = require "fun"
```
Comment (Gerold103) [inline tests/runtest]: One another pure cosmetic change. Please, discard.

File: tests/basic.lua [hunk 2]
```diff
@@ -172,6 +172,48 @@ for _it, a in iter(1, 2, 3, 4, 5, 6, 7) do print(a) end
 error: object 1 of type "number" is not iterable
 --test]]
 
+--------------------------------------------------------------------------------
+-- pairs
+--------------------------------------------------------------------------------
+
+for _it, a in fun.pairs({4, 5, 6}) do print(a) end
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply tests/basic.lua]: Not sure if I really get why somebody would need this fun.pairs().

[the remark this answers is not in the document]
Comment (Gerold103) [reply tests/basic.lua]: It is rather looks like `fun.values`, since it ignores indexes and keys.

File: tests/runtest (hunk 1 above)
Comment (printercu) [reply tests/runtest]: nope. it's assigned to global

File: tests/basic.lua [hunk 3]
```diff
@@ -53,7 +53,7 @@ for _it, a in wrap(wrap(ipairs({}))) do print(a) end
 -- Check that ``iter`` for arrays is equivalent to ``ipairs``
 local t = {1, 2, 3}
 gen1, param1, state1 = iter(t):unwrap()
-gen2, param2, state2 = ipairs(t) 
+gen2, param2, state2 = ipairs(t)
```
[the remark this answers is not in the document]
Comment (printercu) [reply tests/basic.lua]: I use "strip trailing whitespaces" and it's quite complicated to skip all this changes all the time. I can extract this change to the separate commit.

File: tests/basic.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (printercu) [reply tests/basic.lua]: It wraps lua's `pairs` function so it's named the same way. When one read's `fun.pairs` he can easily decide that it works the same way as vanilla `pairs` and has same corner cases. With `values`  he need to read docs to be sure.

It was made similar to all other  fun methods and make this method safe replacement to `fun.iter`, so it provides single argument in chained functions. Maybe it's better to remove current implementation and rename this method to `values` and `map_pairs` to  `pairs`. I like both approaches, we can ask somebody else for advice/suggestion.

**Comment by printercu** [comment]
`iter` can not be fixed in a backward-compatible way. In one case it can produce  key-value iterator, in other - only values. This behaviour is based on `#obj > 0` condition which can be unpredictable in map-tables with numeric keys and sparse arrays.

File: tests/basic.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gerold103) [reply tests/basic.lua]: Sorry, we don't do cosmetic changes just for the sake of them, regardless of text editor settings. The only exception was merge of SQLite into Tarantool, to get rid of SQLite code style. Just drop this hunk.

File: tests/basic.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gerold103) [reply tests/basic.lua]: But it does not work the same as the original pairs. It returns only values. The original pairs returns keys too. `fun.map_pairs` works like the original pairs. `fun.pairs` works like I don't know what. So I still don't understand, why `fun.pairs` is named `pairs`.

Talking of `fun.iter` removal - I don't think we can do that. We will break existing users. If it can't be fixed, then let it be as is.

[the remark this answers is not in the document]
Comment (printercu) [reply tests/basic.lua]: I'm not saying about `iter` removal. It can be neither fixed nor removed in a backward-compatible way.

[the remark this answers is not in the document]
Comment (Gerold103) [reply tests/basic.lua]: Oh, I thought that `Maybe it's better to remove current implementation` was about `fun.iters`. Ok then.

**Comment by rtsisyk** [comment]
Please don't forget to update documentation too!
